### PR TITLE
Fix the WCS bug in MOSViz

### PR DESCRIPTION
**Pull request by ibusko**
This is a temporary, hacky solution, until the entire issue can be handled by the specutils/spectral-cube merge.

This was tested with the MAST files, as well as with other files that have the same structure.

### CODE REVIEW

**Comment by rosteen** [comment]
@ibusko do you have a link to the data I should check this against?

**Comment by ibusko** [comment]
The files are here https://stsci.app.box.com/s/xgxn6587adb8fkjyuvujpyjtwnhtqla0

Note that MOSViz only works if you provide image cutouts. For this data set, I got 4 image cutouts I had from a previous project. For testing purposes, it doesn't matter if they don't correspond exactly to the spectra. It's only their number (4 in this case) that is of significance.

**Comment by javerbukh** [comment]
I downloaded the data from box and am using a cutout from a different dataset. Data is loaded into the table viewer, but when I clicked on it, nothing was loaded at first. It took about 2 minutes of loading before data started appearing in the actual viewers. Each time I clicked on a row, I received the warning snackbar message that the 1D and 2D viewers might get desynced soon. Some 1D spectrum files load as a straight horizontal line or no line at all. I assume this is the data and not the parsing, in which case this PR looks good.

**Review by rosteen — COMMENTED** [review]
I'm also seeing the constant desync panning warning messages - unclear if that's due to issues with this PR or a separate MOSViz fix needed.

The worse problem I'm seeing is that it's very inconsistent what exactly happens when I click a row. There's one file pair that seems to consistently render the 2D and 1D spectra properly, but the others fail most of the time in various ways. Sometimes I get a straight horizontal line for the 1D spectrum, as Jesse said, sometimes I get a vertical line, often with a blank 2D spectrum. It looks like both the 1D and 2D spectra are generally displaying correctly initially but then changing the axes (maybe due to the WCS getting applied?) which messes up the way the data is displayed. It might be worth showing this over screen share, since it's a little hard to describe.

I'm also getting a traceback every time I click a row, but it appears to be due to the dummy 2D image I selected, not a problem with this PR:

```
ValueError                                Traceback (most recent call last)
[clipped for brevity]

~/projects/jdaviz/jdaviz/configs/mosviz/plugins/slit_overlay/slit_overlay.py in place_slit_overlay(self, *args, **kwargs)
     85 
     86             # Use wcs of image viewer to scale slit dimensions correctly
---> 87             wcs_image = WCS(image_data[0].meta)
     88             pixel_region = sky_region.to_pixel(wcs_image)
     89 

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/wcs/wcs.py in __init__(self, header, fobj, key, minerr, relax, naxis, keysel, colsel, fix, translate_units, _do_set)
    414                     header = fits.Header()
    415                     for dict_key in orig_header.keys():
--> 416                         header[dict_key] = orig_header[dict_key]
    417                 except TypeError:
    418                     raise TypeError(

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in __setitem__(self, key, value)
    205             # If we get an IndexError that should be raised; we don't allow
    206             # assignment to non-existing indices
--> 207             self._update((key, value, comment))
    208 
    209     def __delitem__(self, key):

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in _update(self, card)
   1657                 self._modified = True
   1658         elif keyword in Card._commentary_keywords:
-> 1659             cards = self._splitcommentary(keyword, value)
   1660             if keyword in self._keyword_indices:
   1661                 # Append after the last keyword of the same type

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in _splitcommentary(self, keyword, value)
   1899             cards = []
   1900             while idx < len(valuestr):
-> 1901                 cards.append(Card(keyword, valuestr[idx:idx + maxlen]))
   1902                 idx += maxlen
   1903         return cards

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/card.py in __init__(self, keyword, value, comment, **kwargs)
    191                 self.keyword = keyword
    192             if value is not None:
--> 193                 self.value = value
    194 
    195         if comment is not None:

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/card.py in value(self, value)
    328             m = self._ascii_text_re.match(value)
    329             if not m:
--> 330                 raise ValueError(
    331                     'FITS header values must contain standard printable ASCII '
    332                     'characters; {!r} contains characters not representable in '

ValueError: FITS header values must contain standard printable ASCII characters; 'PUTAST: Mar 27 09:00:09 2007 World Coordinate System parameters written\n' contains characters not representable in ASCII or non-printable characters.
```

**Comment by ibusko** [comment]
I haven’t seen these behaviors with my test images. I used four cutouts that I took from another project. Maybe we can’t really use any 2D image for testing; we need images that had worked in MOSViz before? On the other hand, I am not sure that the mapping from gWCS to WCS that I coded is correct.

Sent from my iPhone

On Dec 18, 2020, at 11:51, rosteen <notifications@github.com> wrote:

﻿

@rosteen commented on this pull request.

I'm also seeing the constant desync panning warning messages - unclear if that's due to issues with this PR or a separate MOSViz fix needed.

The worse problem I'm seeing is that it's very inconsistent what exactly happens when I click a row. There's one file pair that seems to consistently render the 2D and 1D spectra properly, but the others fail most of the time in various ways. Sometimes I get a straight horizontal line for the 1D spectrum, as Jesse said, sometimes I get a vertical line, often with a blank 2D spectrum. It looks like both the 1D and 2D spectra are generally displaying correctly initially but then changing the axes (maybe due to the WCS getting applied?) which messes up the way the data is displayed. It might be worth showing this over screen share, since it's a little hard to describe.

I'm also getting a traceback every time I click a row, but it appears to be due to the dummy 2D image I selected, not a problem with this PR:

ValueError                                Traceback (most recent call last)
[clipped for brevity]

~/projects/jdaviz/jdaviz/configs/mosviz/plugins/slit_overlay/slit_overlay.py in place_slit_overlay(self, *args, **kwargs)
     85
     86             # Use wcs of image viewer to scale slit dimensions correctly
---> 87             wcs_image = WCS(image_data[0].meta)
     88             pixel_region = sky_region.to_pixel(wcs_image)
     89

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/wcs/wcs.py in __init__(self, header, fobj, key, minerr, relax, naxis, keysel, colsel, fix, translate_units, _do_set)
    414                     header = fits.Header()
    415                     for dict_key in orig_header.keys():
--> 416                         header[dict_key] = orig_header[dict_key]
    417                 except TypeError:
    418                     raise TypeError(

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in __setitem__(self, key, value)
    205             # If we get an IndexError that should be raised; we don't allow
    206             # assignment to non-existing indices
--> 207             self._update((key, value, comment))
    208
    209     def __delitem__(self, key):

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in _update(self, card)
   1657                 self._modified = True
   1658         elif keyword in Card._commentary_keywords:
-> 1659             cards = self._splitcommentary(keyword, value)
   1660             if keyword in self._keyword_indices:
   1661                 # Append after the last keyword of the same type

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/header.py in _splitcommentary(self, keyword, value)
   1899             cards = []
   1900             while idx < len(valuestr):
-> 1901                 cards.append(Card(keyword, valuestr[idx:idx + maxlen]))
   1902                 idx += maxlen
   1903         return cards

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/card.py in __init__(self, keyword, value, comment, **kwargs)
    191                 self.keyword = keyword
    192             if value is not None:
--> 193                 self.value = value
    194
    195         if comment is not None:

~/opt/anaconda3/envs/viz_dev/lib/python3.8/site-packages/astropy/io/fits/card.py in value(self, value)
    328             m = self._ascii_text_re.match(value)
    329             if not m:
--> 330                 raise ValueError(
    331                     'FITS header values must contain standard printable ASCII '
    332                     'characters; {!r} contains characters not representable in '

ValueError: FITS header values must contain standard printable ASCII characters; 'PUTAST: Mar 27 09:00:09 2007 World Coordinate System parameters written\n' contains characters not representable in ASCII or non-printable characters.

**Review by havok2063 — COMMENTED** [review]
So this sort of works.  When I load the dataset I provided you, it loads ok.  However, each time I click a spectrum row in the table, I get the following error in the snackbar after the dataset loads: `'S_REGION' not found in Spectrum 2D meta attribute`. Syncing between the 1d and 2d views seems to work ok and I get some interactivity.  I do get a warning about panning too far may desync the data viewers when I pan the 1d spectrum profile back and forth.    

When I try loading a different dataset, I get the following traceback when running `load_data`.  `RuntimeError: Keyword SRCTYPE is UNKNOWN.  It should be 'POINT' or 'EXTENDED'. Can't decide between "flux" and "surf_bright" columns.`.  But this could be due to an issue with the test data.  

Some of the data will show a single line in the 1d spectrum profile.  I think this is due to the test files, which aren't necessarily filled with realistic data. 

I think it would be good if we can load these datasets without an associated image cutout.  If the only thing the image is needed/used for is a slit overlay, then that sounds like an optional feature to me.  If it's not guaranteed that nirspec observations will have associated image cutouts, then the cutouts input should be optional.  Otherwise there should be some additional sanity checks on data load to ensure the correct images are supplied along with the 1d/2d spectral files, with some indication of where/how to get the associated images. 

I've added some additional files for testing in the Box here, https://stsci.box.com/s/x1kv642hpoi6y31og65oxt550s8j61ia

**Comment by havok2063** [comment]
@ibusko  any update on this?

**Comment by ibusko** [comment]
I tried with the new test files and I got the same behavior as before. MOSViz seems to work correctly, except for an exception that it raises when first ingesting data (`IndexError: list index out of range` coming from `ipywidgets`). 

@havok2063 the errors you got related to certain FITS keywords seem to be unrelated to the WCS issue we had before. The missing keywords are not part of the WCS definition afaik. I would hazard that the missing `S_REGION` and `SRCTYPE` keywords  where stripped out of the primary header by the same process (in the pipeline) that removes the WCS keywords and adds them back (in the form of an equivalent gWCS) to the ASDF extension trailing the file. But that's just a guess. Anyway, it suggests that the files are broken.

**Comment by havok2063** [comment]
@ibusko Ok, thanks for the update.  I do find it strange why files that used to be loadable are no longer loadable.  Regarding the missing keywords, this is now an issue for other JWST data files outside of MosViz data.  See https://github.com/spacetelescope/jdaviz/issues/390.   The `SRCTYPE` keyword is now missing from all headers.  

@jdavies-st Do you have any idea on why the `SRCTYPE` and `S_REGION` header keywords are now missing from the most recent test JWST data ingested into the Portal?   It looks like the header keyword within the "Target Information" section has changed from `SRCTYPE = 'EXTENDED'           / Advised source type (point/extended)` to `SRCTYAPT= 'UNKNOWN '           / Source type from APT (point/extended)`.   The JWST loaders in specutils contain checks on the `SRCTYPE`.  Do the loaders need to be updated?  Additionally is there any documentation on how the data headers change between pipeline builds?

**Comment by jdavies-st** [comment]
SRCTYPE and SRCTYAPT have been moved out of the PRIMARY header into the SCI or EXTRACT1D extension header.  See

https://github.com/spacetelescope/jwst/pull/4885

**Comment by jdavies-st** [comment]
Not sure about S_REGION.  That has always been in the SCI header, as it should still be.  Feel free to point me to data where this is not the case.  Happy to have a look.

File: jdaviz/configs/mosviz/plugins/parsers.py [hunk 1]
```diff
@@ -125,11 +126,43 @@ def _parse_as_cube(path):
             header['NAXIS3'] = 1
             header['BUNIT'] = 'dN/s'
             header['CUNIT3'] = 'um'
+            header['CTYPE3'] = 'WAVE'
+
+            # Information not present in the SCI header (was removed by
+            # the pipeline) has to be put back so spectral_cube won't
```
Comment (jdavies-st) [inline jdaviz/configs/mosviz/plugins/parsers.py:132]: The FITS WCS was not removed by the pipeline.  It never existed and was never populated to begin with.  The FITS WCS keywords are not generated for any non-IFU spectral data yet.  But we're actually going to be working on this longstanding issue:

https://github.com/spacetelescope/jwst/issues/2685

next sprint, so there might be FITS keywords very soon in `jwst` master.

**Comment by havok2063** [comment]
@jdavies-st Thanks for the update.  I've commented on the `SRCTYPE` issue in https://github.com/spacetelescope/jdaviz/issues/390.   For `S_REGION`, I can confirm that the keyword is in the `SCI` header extension of the s2d file.  There must be a different reason in jdaviz why I am seeing this error.  @ibusko where does this error get generated.  What version of `spectral_cube` are you using?

**Comment by ibusko** [comment]
@havok2063 I can only see that the error is not coming from the jdaviz codebase, but most likely from `spectral_cube`. I am using version 0.5.0.

**Comment by havok2063** [comment]
@ibusko hmm ok, that's the same version I'm using.  I'll dig a little deeper and see if I can reproduce the error.

**Comment by ibusko** [comment]
@havok2063 I can only see that the error is not coming from the jdaviz codebase, but most likely from `spectral_cube`. I am using version 0.5.0.

**Comment by ibusko** [comment]
The `SRCTYPE` keyword issue also affects the ingestion of x1d files by `specviz` (#390). As per discussion in there, this PR should be marked as blocked by https://jira.stsci.edu/browse/JP-1880

**Review by rosteen — APPROVED** [review]
I was able to load the pipeline version 0.17.1 s2d files with these changes, approving

**Review by javerbukh — APPROVED** [review]
Was able to load 0.17.1 data in mosviz, nice work!

**Review by duytnguyendtn — APPROVED** [review]
Here are the results of my testing:

1. The original data we have in the Mosviz Example notebook continues to work and is unaffected by these changes, which is good. No regression as far as I can see
2. With the original data @havok2063 provided (that @ibusko linked: https://stsci.app.box.com/s/xgxn6587adb8fkjyuvujpyjtwnhtqla0), I am seeing the same errors @rosteen and @javerbukh saw, the desyncing errors and inconsistent data loading. Considering the example notebook was unaffected, this sounds like a separate issue
3. With some new data @havok2063 provided (https://stsci.box.com/s/vhzun0fm1uvbfrops717tmhct9uuzdud), dubbed the 0.17.1 data, I was able to successfully load both the 2D and 1D spectra as well, which proves to me this PR works as intended.

Since @javerbukh and I were able to load the 1D spectra in the 0.17.1 data, it would be nice to confirm whether @rosteen is able to load it as well, or if there's another issue we're not seeing. Once Ricky is able to verify the 1D spectra works as well, LGTM!

**Comment by rosteen** [comment]
Well, what the heck, now I'm re-testing and can load the x1d files as well. I swear I didn't change anything (but I tested with specutils release and dev versions in case that was it). Slightly disturbing. 

Anyway, I think this is ok to be merged. I suspect there are perhaps unrelated things going on with Mosviz that will need to be ironed out...

**Pull request merged**